Rewrite the interface between validation and net_processing wrt DoS

TheBlueMatt opened the pull request:
This cleans up an old main-carryover - it made sense that main could decide what DoS scores to assign things because the DoS scores were handled in a different part of main, but now validation is telling net_processing what DoS scores to assign to different things, which is utter nonsense. Instead, we replace CValidationState's nDoS and CorruptionPossible with a general ValidationInvalidReason, which net_processing can handle as it sees fit. I keep the behavior changes here to a minimum, but in the future we can utilize these changes for other smarter behavior, such as disconnecting/preferring to rotate outbound peers based on them providing things which are invalid due to SOFT_FORK because we shouldn't ban for such cases.

This is somewhat complementary with, though obviously conflicts heavily with #11523, which added enums in place of DoS scores, as well as a few other cleanups (which are still relevant).

Compared with previous bans, the following changes are made:
 * Txn with empty vin/vout or null prevouts move from 10 DoS
   points to 100.
 * Loose transactions with a dependency loop now result in a ban
   instead of 10 DoS points.
 * BIP68-violation no longer results in a ban as it is SOFT_FORK.
 * Non-SegWit SigOp violation no longer results in a ban as it
   considers P2SH sigops and is thus SOFT_FORK.
 * Any script violation in a block no longer results in a ban as
   it may be the result of a SOFT_FORK. This should likely be
   fixed in the future by differentiating between them.
 * Proof of work failure moves from 50 DoS points to a ban.
 * Blocks with timestamps under MTP now result in a ban, blocks
   too far in the future continue to *not* result in a ban.
 * Inclusion of non-final transactions in a block now results in a
   ban instead of 10 DoS points.

TheBlueMatt committed:
Add useful-for-dos "reason" field to CValidationState

This is a first step towards cleaning up our DoS interface - make
validation return *why* something is invalid, and let net_processing
figure out what that implies in terms of banning/disconnection/etc.

TheBlueMatt committed:
Add functions to convert CValidationInterface's reason to DoS info

CValidationInterface's new GetReason() field provides more than
enough information for net_processing to determine what to do with
the peer, use that instead of trusting validation's idea of what
should result in a DoS ban.

Compared with previous bans, the following changes are made:
 * Txn with empty vin/vout or null prevouts move from 10 DoS
   points to 100.
 * Loose transactions with a dependency loop now result in a ban
   instead of 10 DoS points.
 * Any pre-segwit soft-fork errors (ie all soft-fork errors)
   now result in a ban.
 * Proof of work failure moves from 50 DoS points to a ban.
 * Blocks with timestamps under MTP now result in a ban, blocks
   too far in the future continue to *not* result in a ban.
 * Inclusion of non-final transactions in a block now results in a
   ban instead of 10 DoS points.

TheBlueMatt committed:
Prep for scripted-diff by removing some \ns which annoy sed.

TheBlueMatt committed:
Use state reason field to check for collisions in cmpctblocks

TheBlueMatt committed:
Remove references to CValidationState's DoS and CorruptionPossible

TheBlueMatt commented:
Rebased.

promag commented on src/test/txvalidation_tests.cpp:
```diff
@@ -53,9 +53,8 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_reject_coinbase, TestChain100Setup)
-    BOOST_CHECK_EQUAL(state.IsInvalid(nDoS), true);
-    BOOST_CHECK_EQUAL(nDoS, 100);
+    BOOST_CHECK_EQUAL(state.IsInvalid(), true);
+    BOOST_CHECK(state.GetReason() == ValidationInvalidReason::CONSENSUS);
```
Also BOOST_CHECK_EQUAL?

TheBlueMatt replied:
Doesn't compile as GetReason() can't be printed.

Empact replied:
Not worthy of revisiting here, but fyi I added a facility here to support `BOOST_CHECK_EQUAL` for enum: https://github.com/bitcoin/bitcoin/pull/13076/files#diff-4a0cb5ad5e93d187a1b065a99bbae143

promag commented on src/validation.cpp:
```diff
@@ -1906,11 +1907,15 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
             if (!Consensus::CheckTxInputs(tx, state, view, pindex->nHeight, txfee)) {
+                // CheckTxInputs may return MISSING_INPUTS but we can't return that, as
+                // its nonsense for a block, so we reset the reason flag to CONSENSUS here.
+                state.Invalid(ValidationInvalidReason::CONSENSUS, false,
```
Make it `CONSENSUS` only if `NOT_STANDARD`? I know that atm only those 2 are set but, imo, it's more clear to do so inside that condition:
```cpp
if (state.GetReason() == ValidationInvalidReason::NOT_STANDARD) {
    state.Invalid(ValidationInvalidReason::CONSENSUS, ...);
}
```
The same happens in line 1951.

TheBlueMatt replied:
Done.

promag replied:
ATM `CheckTxInputs` set `MISSING_INPUTS` and `CONSENSUS` reasons and only makes sense to override when reason is `MISSING_INPUTS`:
```cpp
if (state.GetReason() == ValidationInvalidReason::MISSING_INPUTS) {
    state.Invalid(ValidationInvalidReason::CONSENSUS, ...);
}
```

TheBlueMatt replied:
Errrr....fixed, actually this time.

promag reviewed (commented):
Concept ACK.

Do you think it's reasonable to treat `ValidationInvalidReason` like `CONSENSUS | MISSING_INPUTS`, so you could accumulate reasons instead of overwriting?

promag commented on src/consensus/validation.h:
```diff
@@ -22,6 +22,32 @@ static const unsigned char REJECT_NONSTANDARD = 0x40;
+  */
+enum class ValidationInvalidReason {
+    // txn and blocks:
+    VALID,           //!< not actually invalid
```
`ValidationInvalidReason::NONE` instead?

TheBlueMatt replied:
Sure, sounds good.

sdaftuar commented on src/validation.cpp:
```diff
@@ -1906,11 +1914,17 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
+                    // CheckTxInputs may return MISSING_INPUTS but we can't return that, as
+                    // its nonsense for a block, so we reset the reason flag to CONSENSUS here.
+                    state.DoS(state.GetDoS(), ValidationInvalidReason::CONSENSUS, false,
+                              state.GetRejectCode(), state.GetRejectReason(), state.GetDebugMessage());
```
The arguments don't match here:
```
validation.cpp:1921:79: error: no viable conversion from 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char> >') to 'bool'
                              state.GetRejectCode(), state.GetRejectReason(), state.GetDebugMessage());
```
but more generally I think this would be a lot more readable if we could just add an explicit way to set the invalid reason on a CValidationState, rather than invoke DoS() in this way (here and elsewhere as well).

sipa replied:
This seems fixed now, @sdaftuar?

sdaftuar commented on src/validation.cpp:
```diff
@@ -1866,7 +1874,7 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
             for (size_t o = 0; o < tx->vout.size(); o++) {
                 if (view.HaveCoin(COutPoint(tx->GetHash(), o))) {
-                    return state.DoS(100, error("ConnectBlock(): tried to overwrite transaction"),
+                    return state.DoS(100, ValidationInvalidReason::SOFT_FORK, error("ConnectBlock(): tried to overwrite transaction"),
```
I think this should probably be CONSENSUS rather than SOFT_FORK.

TheBlueMatt replied:
We should figure out where we want to draw a line between CONSENSUS (ban the peer) and SOFT_FORK (ignore the peer). Obviously this case *is* a SOFT_FORK, but at this point its so old that it may not be worth keeping that distinction. I aired on the side of "let any node, no matter how old, not get banned" in general.

sdaftuar commented on src/validation.cpp:
```diff
@@ -3136,7 +3164,7 @@ static bool ContextualCheckBlock(const CBlock& block, CValidationState& state, c
     for (const auto& tx : block.vtx) {
         if (!IsFinalTx(*tx, nHeight, nLockTimeCutoff)) {
-            return state.DoS(10, false, REJECT_INVALID, "bad-txns-nonfinal", false, "non-final transaction");
+            return state.DoS(10, ValidationInvalidReason::CONSENSUS, false, REJECT_INVALID, "bad-txns-nonfinal", false, "non-final transaction");
```
Note that this could be "SOFT_FORK" (BIP 113).  But as I wrote elsewhere I think we should just use CONSENSUS everywhere in place of SOFT_FORK, for now.

sdaftuar commented on src/validation.cpp:
```diff
@@ -3194,7 +3222,9 @@ static bool ContextualCheckBlock(const CBlock& block, CValidationState& state, c
-        return state.DoS(100, false, REJECT_INVALID, "bad-blk-weight", false, strprintf("%s : weight limit failed", __func__));
+        // We can call this a consensus failure as any data-providers who provided
+        // us with witness data can be expected to support SegWit validation.
+        return state.DoS(100, ValidationInvalidReason::CONSENSUS, false, REJECT_INVALID, "bad-blk-weight", false, strprintf("%s : weight limit failed", __func__));
```
If we're going to use this line of reasoning, then I think we may as well also not use SOFT_FORK anywhere else either (at least for invalid blocks, perhaps loose transactions should be different) -- presumably no nodes that support segwit validation should be unaware of any of the prior soft forks.

TheBlueMatt replied:
Huh? Doing so would break the ability of a pre-segwit node to provide us a non-segwit-but-valid block. Further, this line of reasoning does *not* assume that we only download blocks from segwit-enabled peers (which may be the case in practice in our current net layer, but making that assumption strong in validation seems bad). The line of reasoning here is only applicable to nodes that *provided* the witness, not nodes that (for some reason) does not.

sdaftuar commented on src/validation.cpp:
```diff
@@ -1934,17 +1948,25 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
+                    // CheckInputs may return NOT_STANDARD for extra flags we passed,
+                    // but we can't return that, as its nonsense, so we reset the
+                    // reason flag to SOFT_FORK here.
+                    state.DoS(state.GetDoS(), ValidationInvalidReason::SOFT_FORK, false,
```
1) As all parallel-validation script failures are SOFT_FORK further down, I think we ought to make the non-parallel validation case be the same, rather than filter for NOT_STANDARD first.
2) Since our blocks are only coming from segwit peers, I actually think we should just change all these to CONSENSUS anyway, and leave a comment explaining that when we deploy a new soft fork, we need to be careful about not splitting the network.

TheBlueMatt replied:
I'm confused. The NOT_STANDARD return here is to keep behavior the same for transactions as well as blocks. NOT_STANDARD is not a valid response for block ValidationReasons, so we have to swap it here.

sdaftuar replied:
Re 1), I'm just saying that we should always rewrite the reason to be the same as we'd get if we did parallel validation (which is SOFT_FORK in this PR currently), versus only return SOFT_FORK if the reason was NOT_STANDARD.  It doesn't really make sense for ConnectBlock() to behave differently with regard to error codes if using single vs parallel script checking.

sdaftuar commented on src/validation.cpp:
```diff
@@ -3168,11 +3189,11 @@ static bool ContextualCheckBlock(const CBlock& block, CValidationState& state, c
             // witness tree.
             if (block.vtx[0]->vin[0].scriptWitness.stack.size() != 1 || block.vtx[0]->vin[0].scriptWitness.stack[0].size() != 32) {
-                return state.DoS(100, false, REJECT_INVALID, "bad-witness-nonce-size", true, strprintf("%s : invalid witness nonce size", __func__));
+                return state.Invalid(ValidationInvalidReason::MUTATED, false, REJECT_INVALID, "bad-witness-nonce-size", strprintf("%s : invalid witness nonce size", __func__));
```
Not really obvious here why we use MUTATED rather than WITNESS_MUTATED, here and in the next few lines (other than needing to match the code in InvalidBlockFound() and AcceptBlock(), which only check against MUTATED). Perhaps add a comment explaining the usage somewhere?

TheBlueMatt replied:
WITNESS_MUTATED is only for transactions. In the future we could separate out the InvalidReason into two different ones for transactions/blocks.

sdaftuar reviewed (changes requested):
Concept ACK, but don't think we need SOFT_FORK to be used anywhere now.

TheBlueMatt commented:
Commented on the points I disagreed with, I obviously strongly disagree about making assumptions about the net_processing layer in validation, as that's been one of my overarching goals over the past few years. Will address the other points when I get a chance to rebase.

sdaftuar commented:
The more I think about this, the more I think we should drop the distinction between "SOFT_FORK" and "CONSENSUS" altogether.  At least when it comes to block processing, the only concept that matters is CONSENSUS.

The reason we care to distinguish SOFT_FORK from CONSENSUS is to highlight recent rule changes that our peers may not know about, so that the net_processing layer can optionally do something different that for SOFT_FORK violations than it would do for CONSENSUS violations.  However I don't think this is a terribly meaningful distinction.  The main goals of the net_processing/net layers, with respect to these validation issues, is:
 * keep us connected to the honest network
 * prevent attackers from DoS'ing us
 * don't needlessly partition old nodes that have fallen back to SPV security because they're unaware of the consensus rules

So what is our strategy for each of these points?  For keeping us connected to the honest network, I think we mainly rely on having robust outbound peers -- any outbound peer that is not enforcing all the consensus rules should be disconnected.  Combined with the mitigations we deployed last fall for outbound peers whose tips aren't keeping up, and for protection in the case that our tip hasn't advanced in a while, I believe this is sufficient.  (See https://gist.github.com/sdaftuar/c2a3320c751efb078a7c1fd834036cb0 for more details around my thinking here.)

For not needlessly partitioning old nodes -- I think it's sufficient to not disconnect inbound peers that relay blocks that violate consensus rules, regardless of whether the rule is a recent soft fork or not.

And to prevent DoS -- I think we just ensure that any blocks have valid proof-of-work, combined with protections against processing blocks that are too far behind our tip (or on a too-little-work-chain).

In all of these cases, I don't think we need to distinguish "SOFT_FORK" from "CONSENSUS".  And as its pretty arbitrary about when to decide to move something from the SOFT_FORK designation to CONSENSUS, I think we may as well not bother.

TheBlueMatt commented:
Discussed this offline with @sdaftuar for a bit - I'm OK with redefining "SOFT_FORK" to mean "any future SOFT_FORK after segwit, so is currently unused", but I'd rather leave it there as we'd certainly need it for "any future SOFT_FORK". Happy to do so if others like. Still want more review here, though.

sdaftuar commented:
No objection to leaving "SOFT_FORK" in for future use (even if I'm not sure we'll ever use it, I am open to being convinced).  Nit: I would somewhat prefer to call it something other than "SOFT_FORK" -- perhaps "RECENT_CONSENSUS_CHANGE" or something else that indicates the relevant substance rather than the form?

sipa commented:
Big concept ACK, I'm happy to see the DoS scores get out of the network processing logic.

TheBlueMatt commented:
@sipa well, current master consider RECENT_CONSENSUS_CHANGE/not-banning as any soft fork, ever, kinda, but also some other stuff, but not SegWit things. Its kinda a grab-bag as @sdaftuar points out. This PR cleaned it up somewhat but aired on the side of "all soft-forks", @sdaftuar wants it to air more on the side of "things after segwit", which I think makes sense. Leaving it the way it is on master kinda makes no sense (and would be hard to do).

sdaftuar commented:
In the interests of advancing this refactor, I'm fine with doing whatever is least controversial (presumably, less change to existing behavior) in this PR, and saving the larger behavioral changes for a future PR.

TheBlueMatt committed:
Use new reason-based DoS/disconnect logic instead of state.nDoS

TheBlueMatt commented:
Rebased and changed SOFT_FORK to RECENT_CONSENSUS_CHANGE, redefining it to mean "change more recent than segwit" (ie it is currently unused, as @sdaftuar requested).

skeees commented:
Concept ack to the motivation - while validation shouldn't be deciding how many points to assign peers, you could argue that it should be telling net processing "this is rejected because I don't have enough information to assess that it is valid" vs. "this is invalid because it was maliciously constructed" - it would be nice from a readability perspective if that were more immediately apparent (e.g. by categorizing the reason names or the like)

TheBlueMatt commented:
I think the only cases where we can say "I'm missing something this is based on" is ValidationInvalidReason::MISSING_PREV and MISSING_INPUTS, which are pretty self-explanitory. Open to suggestions if you had something else in mind (or a comment to clarify).

MarcoFalke commented on src/validation.cpp:
```diff
@@ -3065,28 +3084,30 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
-                                 strprintf("Transaction check failed (tx hash %s) %s", tx->GetHash().ToString(), state.GetDebugMessage()));
+    for (const auto& tx : block.vtx) {
+        if (!CheckTransaction(*tx, state, false)) {
+            LogPrintf("Transaction check failed (tx hash %s) %s", tx->GetHash().ToString(), state.GetDebugMessage());
```
from the linter: "All calls to LogPrintf() should be terminated with \n"

jnewbery commented on src/consensus/validation.h:
```diff
@@ -71,19 +96,7 @@ class CValidationState {
-        }
-        return false;
-    }
-    bool CorruptionPossible() const {
```
Alternatively, you could just change this function to return whether `reason` is set to `MUTATED` or `WITNESS_MUTATED`

TheBlueMatt replied:
Then it'd be dead code and forcing the caller to be more clear about what *exactly* they're checking for is much more stable.

jnewbery replied:
Understood. This is fine as is.

jnewbery commented:
I've left a bunch of comments, mostly around making the commenting clearer.

Empact commented on src/consensus/validation.h:
```diff
@@ -22,6 +22,39 @@ static const unsigned char REJECT_NONSTANDARD = 0x40;
+  * These are much more granular than the rejection codes, which may be more
+  * useful for some other use-cases.
+  */
+enum class ValidationInvalidReason {
```
Renaming this `InvalidReason` would save a lot of chars overall. The only other prefix in this module is `REJECT_`

Empact commented on src/consensus/validation.h:
```diff
@@ -22,6 +22,39 @@ static const unsigned char REJECT_NONSTANDARD = 0x40;
+    RECENT_CONSENSUS_CHANGE,
+    UNKNOWN_INVALID,         //!< this object was cached as being invalid, but we don't know why
+    // Only blocks:
+    MUTATED,                 //!< the block's data didn't match the data committed to by the PoW
```
Agreed

TheBlueMatt replied:
I'd rather just split it up into separate validation state objects in a later refactor, then its type-checked  instead.

jnewbery replied:
Sounds good, but is there any downside to making it clearer until that future refactor?

Empact commented on src/validation.cpp:
```diff
@@ -3065,28 +3084,30 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
-                                 strprintf("Transaction check failed (tx hash %s) %s", tx->GetHash().ToString(), state.GetDebugMessage()));
+    for (const auto& tx : block.vtx) {
+        if (!CheckTransaction(*tx, state, false)) {
+            LogPrintf("Transaction check failed (tx hash %s) %s\n", tx->GetHash().ToString(), state.GetDebugMessage());
```
The `Invalid` call here just adds to the debug info - see that each arg is carried over from the existing state, and the debug `strprintf` prepends a message to the existing debug info. Logging is potentially an equivalently communicative alternative, though I would use `FormatStateMessage` here to include all the state info.

Empact replied:
nit: I still think the full `FormatStateMessage` is the better info to record here.

Empact commented:
Is it possible to also kill the `REJECT_*` codes here? They seem possibly duplicative with the new reason value.

TheBlueMatt committed:
Update some comments in validation.cpp as we arent doing DoS there

TheBlueMatt commented on src/consensus/validation.h:
```diff
@@ -22,6 +22,39 @@ static const unsigned char REJECT_NONSTANDARD = 0x40;
 static const unsigned char REJECT_INSUFFICIENTFEE = 0x42;
 static const unsigned char REJECT_CHECKPOINT = 0x43;
 
+/** A "reason" why something was invalid, suitable for (possibly) getting angry
```
Gotta ruin all my fun...fixed.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -560,28 +560,31 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
     if (!gArgs.GetBoolArg("-prematurewitness", false) && tx.HasWitness() && !witnessEnabled) {
-        return state.DoS(0, false, REJECT_NONSTANDARD, "no-witness-yet", true);
+        // We call this WITNESS_MUTATED because we don't bother to check if the witness was maybe garbage
+        // and its important clients realize that they shouldn't assume the witness indicated the
```
Removed the comment and replaced it with a more detailed comment on the WITNESS_MUTATED definition.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -710,21 +713,21 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         // merely non-standard transaction.
         if (nSigOpsCost > MAX_STANDARD_TX_SIGOPS_COST)
-            return state.DoS(0, false, REJECT_NONSTANDARD, "bad-txns-too-many-sigops", false,
+            return state.Invalid(ValidationInvalidReason::NOT_STANDARD, false, REJECT_NONSTANDARD, "bad-txns-too-many-sigops",
```
Removed the comment, looks like it was just wrong.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1422,7 +1423,7 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
                     // peering with non-upgraded nodes even after soft-fork
                     // super-majority signaling has occurred.
-                    return state.DoS(100,false, REJECT_INVALID, strprintf("mandatory-script-verify-flag-failed (%s)", ScriptErrorString(check.GetScriptError())));
+                    return state.Invalid(ValidationInvalidReason::CONSENSUS, false, REJECT_INVALID, strprintf("mandatory-script-verify-flag-failed (%s)", ScriptErrorString(check.GetScriptError())));
```
Removed the comment. It isnt adding anything that the RECENT_CONSENSUS_CHANGE documentation didnt already cover.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2024,13 +2039,17 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
                                block.vtx[0]->GetValueOut(), blockReward),
                                REJECT_INVALID, "bad-cb-amount");
 
+    // TODO: We should be able to ban peers that send us completely bogus scripts,
```
Ah missed removing that when I changed things for @sdaftuar's suggestion. Removed.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1971,11 +1972,17 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
             if (!Consensus::CheckTxInputs(tx, state, view, pindex->nHeight, txfee)) {
+                if (state.GetReason() == ValidationInvalidReason::MISSING_INPUTS) {
+                    // CheckTxInputs may return MISSING_INPUTS but we can't return that, as
+                    // its nonsense for a block, so we reset the reason flag to CONSENSUS here.
```
Fixed.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -784,6 +784,75 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message)
+    case ValidationInvalidReason::INVALID_PREV:
+    case ValidationInvalidReason::MISSING_PREV:
+        return true;
+    // Speed-of-light or out-of-sync:
```
Clearer?

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -784,6 +784,75 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message)
+    // The node is mean:
+    case ValidationInvalidReason::CONSENSUS:
+    case ValidationInvalidReason::MUTATED:
+    case ValidationInvalidReason::UNKNOWN_INVALID: // Really "duplicate of invalid "
```
The enum definition is pretty clear IMO "this object was cached as being invalid, but we don't know why". Any suggestions for a different comment?

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -956,14 +1025,12 @@ void PeerLogicValidation::BlockChecked(const CBlock& block, const CValidationSta
             State(it->second.first)->rejects.push_back(reject);
-            if (nDoS > 0 && it->second.second)
-                Misbehaving(it->second.first, nDoS);
+            MaybePunishNode(it->second.first, state, !it->second.second);
```
Happy to do it in a followup, this is already enough code and I don't want to risk a bug if it sits for another 6 months and requires rebase.

jnewbery replied:
Yes - can be done in a follow up, but in the meantime:

` MaybePunishNode(it->second.first, state, !it->second.second);`

is almost completely inscrutable!

Empact replied:
In fairness, it was this way when he found it. :P Def worthy of cleanup.

Empact replied:
FYI, a more self-documenting alternative looks something like:
```c++
if (state.IsInvalid()) {      
    // Don't send reject message with code 0 or an internal reject code.
    NodeId node_id;
    bool punish_if_block_invalid;
    std::tie(node_id, punish_if_block_invalid) = it->second;
    if (it != mapBlockSource.end() && State(node_id) && state.GetRejectCode() > 0 && state.GetRejectCode() < REJECT_INTERNAL) {
        CBlockReject reject = {(unsigned char)state.GetRejectCode(), state.GetRejectReason().substr(0, MAX_REJECT_MESSAGE_LENGTH), hash};
        State(node_id)->rejects.push_back(reject);
        MaybePunishNode(node_id, state, !punish_if_block_invalid);
    }
}
```
Note you can early return for the `== end()` case here because it's present in every case.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -1406,6 +1466,10 @@ bool static ProcessHeadersMessage(CNode *pfrom, CConnman *connman, const std::ve
                 pfrom->fDisconnect = true;
             }
+            if (!MaybePunishNode(pfrom->GetId(), state, false, "invalid header received")) {
+                // MaybePunishNode returning true indicates Misbehaving logged the error for us
```
Done.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -784,6 +784,75 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message)
+    // Speed-of-light or out-of-sync:
+    case ValidationInvalidReason::MISSING_PREV:
+        {
+            // TODO: Handle this much more gracefully
```
Fixed.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -784,6 +784,77 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message)
+    // The node is is providing invalid data:
+    case ValidationInvalidReason::CONSENSUS:
+    case ValidationInvalidReason::MUTATED:
+    case ValidationInvalidReason::UNKNOWN_INVALID: // Really "duplicate of invalid "
```
"where the enum is defined" - I'm confused, at the enum definition it says "this object was cached as being invalid, but we don't know why", is that insufficient?

jnewbery replied:
Apparently it is insufficient, since you've felt the need to leave a comment here explaining what you really mean (and repeated it in the function below) :slightly_smiling_face:

Empact replied:
Agreed the comment should be carried over, or the value renamed.

TheBlueMatt replied:
Cleaned up.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2304,8 +2370,7 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 // case.
-                int nDoS = 0;
-                if (!state.IsInvalid(nDoS) || nDoS == 0) {
+                if (!state.IsInvalid() || !MayResultInDisconnect(state, false)) {
```
At this point, it's clearer to switch the ordering of the conditionals:

```
                if (state.IsInvalid() && MayResultInDisconnect(state, false)) {
                    LogPrintf("Not relaying invalid transaction %s from whitelisted peer=%d (%s)\n", tx.GetHash().ToString(), pfrom->GetId(), FormatStateMessage(state));
                } else {
                    LogPrintf("Force relaying tx %s from whitelisted peer=%d\n", tx.GetHash().ToString(), pfrom->GetId());
                    RelayTransaction(tx, connman);
                }
```

if (THING AND THING) is more natural than if (^THING OR ^THING)

TheBlueMatt replied:
Done.

jnewbery reviewed (commented):
Does it make sense to squash the commits _Use state reason field to check for collisions in cmpctblocks_ and _Remove references to CValidationState's DoS and CorruptionPossible_?

jnewbery commented:
Much clearer to me now. Thanks for the updates!

A few more comments inline.

Empact commented:
Renaming `ValidationInvalidReason` to `InvalidReason` would save 1,130 chars. Think of all those chars. ;)

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -748,7 +743,7 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
             if (setConflicts.count(hashAncestor))
             {
-                return state.DoS(10, false,
+                return state.DoS(10, ValidationInvalidReason::CONSENSUS, false,
```
MISSING_INPUTS and CONFLICT imply that its possible the transaction may be valid, but we're missing something or otherwise unable to add it to the mempool right now. That is not the case here - there is no way this transaction could have been included in the mempool of any well-behaved node, nor could it ever make it into a block, so CONSENSUS is correct here.

sipa replied:
Ah, I see. The block "missing inputs" case is also reported as CONSENSUS; makes sense.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1999,17 +2005,25 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
-            if (!CheckInputs(tx, state, view, fScriptChecks, flags, fCacheResults, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : nullptr))
+            if (!CheckInputs(tx, state, view, fScriptChecks, flags, fCacheResults, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : nullptr)) {
+                if (state.GetReason() == ValidationInvalidReason::NOT_STANDARD) {
+                    // CheckInputs may return NOT_STANDARD for extra flags we passed,
```
In an earlier version of the patch it may have returned NON_STANDARD for flags which were added in a soft fork, but at @sdaftuar's suggestion that is no longer the case. Will fix the comment, but we'll need this when we need RECENT_CONSENSUS_CHANGE at some point.

TheBlueMatt commented on src/consensus/validation.h:
```diff
@@ -22,6 +22,43 @@ static const unsigned char REJECT_NONSTANDARD = 0x40;
+     * non-standard witnesses).
+     */
+    WITNESS_MUTATED,
+    CONFLICT,                //!< tx already in mempool or conflicts with an existing one (which is in chain or in mempool and RBF failed)
```
Seems reasonable. Will do.

MarcoFalke commented:
Needs rebase. (Presumably due to #13032)

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1971,11 +1971,17 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
             if (!Consensus::CheckTxInputs(tx, state, view, pindex->nHeight, txfee)) {
+                if (state.GetReason() == ValidationInvalidReason::MISSING_INPUTS) {
+                    // CheckTxInputs may return MISSING_INPUTS but we can't return that, as
+                    // its not defined for a block, so we reset the reason flag to CONSENSUS here.
```
Fixed

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1999,17 +2005,25 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
+            if (!CheckInputs(tx, state, view, fScriptChecks, flags, fCacheResults, fCacheResults, txdata[i], nScriptCheckThreads ? &vChecks : nullptr)) {
+                if (state.GetReason() == ValidationInvalidReason::NOT_STANDARD) {
+                    // CheckInputs may return NOT_STANDARD for extra flags we passed,
+                    // but we can't return that, as its not defined for a block, so
```
Fixed

TheBlueMatt commented:
Rebased and addressed all outstanding feedback (I believe). I didnt opt for shortening the name of "ValidationInvalidReason"...I dont think we should be shortening unless things are just as clear as the more verbose name.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2228,19 +2295,18 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                         LogPrint(BCLog::MEMPOOL, "   removed orphan tx %s\n", orphanHash.ToString());
                         vEraseQueue.push_back(orphanHash);
-                        if (!orphanTx.HasWitness() && !stateDummy.CorruptionPossible()) {
+                        if (!orphanTx.HasWitness() && state.GetReason() != ValidationInvalidReason::TX_WITNESS_MUTATED) {
```
Should this be `orphan_state` instead of `state`?

sipa commented:
utACK f29c11a45edb138711ae7004b3b550ba416fc215

MarcoFalke commented:
weak utACK f29c11a45edb138711ae7004b3b550ba416fc215 (Did a single pass over all commits)

sipa commented:
Needs rebase due to #13185.

TheBlueMatt commented:
(trivially) rebased. Would love another once-over from @sdaftuar.

TheBlueMatt committed:
scripted-diff: Remove DoS calls to CValidationState

-BEGIN VERIFY SCRIPT-
sed -i 's/\.DoS(\(.*\), REJECT_\(.*\), \(true\|false\)/.DoS(\1, REJECT_\2/' src/validation.cpp src/consensus/tx_verify.cpp
sed -i 's/state.GetRejectCode(), state.GetRejectReason(), state.CorruptionPossible(), state.GetDebugMessage())/state.GetRejectCode(), state.GetRejectReason(), state.GetDebugMessage())/' src/validation.cpp
sed -i 's/\.DoS([^,]*, /.Invalid\(/' src/validation.cpp src/consensus/tx_verify.cpp
-END VERIFY SCRIPT-

sipa commented:
re-utACK e73cf5195daa5b740775a1cd07ece992c6a89cae (verified that the only diff since my previous ACK after a rebase is dealing with #11423 and #13185).

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -1485,6 +1476,7 @@ bool static ProcessHeadersMessage(CNode *pfrom, CConnman *connman, const std::ve
                 // etc), and not just the duplicate-invalid case.
                 pfrom->fDisconnect = true;
             }
+            MaybePunishNode(pfrom->GetId(), state, false, "invalid header received");
```
I believe this can result in a behavior change, where we would DoS-ban peers that send duplicate headers that are invalid.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -2440,14 +2427,8 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
-                    LogPrint(BCLog::NET, "Peer %d sent us invalid header via cmpctblock\n", pfrom->GetId());
-                }
+            if (state.IsInvalid()) {
+                MaybePunishNode(pfrom->GetId(), state, true, "invalid header via cmpctblock");
```
This also appears to be a behavior change; some headers failures currently result in a Misbehaving call, but as-is I believe compact block peers are never punished for errors here?

sdaftuar reviewed (changes requested):
Looks pretty good; left some nits and spotted a couple behavior changes with headers processing that should be addressed.
